### PR TITLE
Fan check

### PR DESCRIPTION
**Pull request by Hagridpon3**
Add support for fan check. This feature is enabled via settings menu.

### CODE REVIEW

File: src/common/marlin_server.cpp [hunk 1]
```diff
@@ -581,6 +589,61 @@ static void _server_print_loop(void) {
...
+            set_warning(WarningType::PrintFanError);
+            if (marlin_server.print_state == mpsPrinting) {
+                marlin_server.print_state = mpsPausing_Begin;
+            }
+            print_fan_error = 0;
+            error_dialog_show = true;
+            _dbg("print fan not spin dialog show!");
+            osDelay(50);
+        }
+    }
```
Comment (dragomirecky) [inline src/common/marlin_server.cpp:636]: As mentioned above, this is just too much of logic for marlin_server. I believe it shouldn't be here. It fits the best to Marlin itself, but we don't have fan monitoring there, so the second-best option seems to create a separate module in our firmware.

File: src/gui/dialogs/window_dlg_strong_warning.cpp [hunk 2]
```diff
@@ -12,6 +13,7 @@ window_dlg_strong_warning_t::window_dlg_strong_warning_t()
 
 void window_dlg_strong_warning_t::Show(string_view_utf8 txt) {
     static window_dlg_strong_warning_t dlg;
+    Sound_Play(eSOUND_TYPE::StandardAlert);
```
Comment (dragomirecky) [inline src/gui/dialogs/window_dlg_strong_warning.cpp:16]: Are you sure this is what you want? To always play an alert sound for every warning dialog?

File: src/gui/guimain.cpp [hunk 3]
```diff
@@ -80,8 +80,8 @@ void MsgCircleBuffer_cb(const char *txt) {
 }
 
 void Warning_cb(WarningType type) {
-    static constexpr const char *HotendFanErrorMsg = N_("I am dummy HotendFanErrorMsg, I need to be replaced with something else ... ");
-    static constexpr const char *PrintFanErrorMsg = N_("I am dummy PrintFanError, I need to be replaced with something else ... ");
+    static constexpr const char *HotendFanErrorMsg = N_("Err: EXTR. FAN ERROR");
+    static constexpr const char *PrintFanErrorMsg = N_("Err: PRINT FAN ERROR");
```
Comment (dragomirecky) [inline src/gui/guimain.cpp:84]: The "error" world is there two times.

What about something more user friendly, like:
```C++
static constexpr const char *HotendFanErrorMsg = N_("Extruder fan error");
static constexpr const char *PrintFanErrorMsg = N_("Print fan error");
```

Comment (Hagridpon3) [reply src/gui/guimain.cpp]: Solved

File: src/gui/ScreenHandler.hpp [hunk 4]
```diff
@@ -58,6 +58,10 @@ class Screens {
     void DisableMenuTimeout();
     bool GetMenuTimeout();
 
+    void EnableFanCheck();
+    void DisableFanCheck();
+    bool GetFanChceck();
```
[the remark this answers is not in the document]
Comment (Hagridpon3) [reply src/gui/ScreenHandler.hpp:63]: Solved.

File: src/common/eeprom.cpp [hunk 5]
```diff
@@ -196,6 +198,7 @@ static const eeprom_vars_t eeprom_var_defaults = {
     {"Custom3", FLT_MAX },
     {"Custom4", FLT_MAX },
 	0,               // EEVAR_SELFTEST_RESULT
+    0,
```
[the remark this answers is not in the document]
Comment (Hagridpon3) [reply src/common/eeprom.cpp]: Solved

File: src/common/marlin_server.cpp [hunk 6]
```diff
@@ -107,6 +107,14 @@ uint32_t *pCommand = &marlin_server.command;
 #endif
 marlin_server_idle_t *marlin_server_idle_cb = 0; // idle callback
 
+uint32_t fan_time = 0;
+uint8_t nozzle_fan_error = 0;
+uint8_t print_fan_error = 0;
+bool error_dialog_show = false;
+uint8_t minimal_fan_rpm_for_detect_spining = 50;
+uint16_t scan_fan_time = 1500;
+bool fan_check_enable = false;
```
[the remark this answers is not in the document]
Comment (Hagridpon3) [reply src/common/marlin_server.cpp]: Solved

File: src/common/marlin_server.cpp (hunk 1 above)
Comment (Hagridpon3) [reply src/common/marlin_server.cpp]: Solved

File: src/gui/dialogs/window_dlg_strong_warning.cpp (hunk 2 above)
Comment (Hagridpon3) [reply src/gui/dialogs/window_dlg_strong_warning.cpp]: Play StandardAlert only for fan error.

File: src/common/eeprom.cpp [hunk 7]
```diff
@@ -91,6 +91,7 @@ typedef struct _eeprom_vars_t {
     Sheet SHEET_PROFILE7;
     uint32_t SELFTEST_RESULT;
     uint8_t QR_PRIVACY;
+    uint8_t FAN_CHECK;
```
Comment (dragomirecky) [inline src/common/eeprom.cpp:94]: /suggestion/ Consider renaming it to `FAN_CHECK_ENABLED` to make It clear that it stores a boolean

File: src/common/marlin_server.cpp [hunk 8]
```diff
@@ -248,6 +250,7 @@ int marlin_server_cycle(void) {
     if ((tick - marlin_server.last_update) > MARLIN_UPDATE_PERIOD) {
         marlin_server.last_update = tick;
         changes = _server_update_vars(marlin_server.update_vars);
+        //marlin_server.vars.fan_check = variant_get_ui8(eeprom_get_var(EEVAR_FAN_CHECK));
```
Comment (dragomirecky) [inline src/common/marlin_server.cpp:253]: Something to remove?

File: src/common/marlin_server.cpp [hunk 9]
```diff
@@ -502,12 +506,13 @@ static void _server_print_loop(void) {
         gcode.reset_stepper_timeout(); //prevent disable axis
         break;
     case mpsResuming_Begin:
-        if (marlin_server_print_reheat_ready()) {
-            marlin_server_unpark_head();
-            marlin_server.print_state = mpsResuming_UnparkHead;
+        if (marlin_server.vars.fan_check) {
+            if (fanctl1.getState() == CFanCtl::running) {
```
Comment (dragomirecky) [inline src/common/marlin_server.cpp:510]: /question/ Not sure I follow the logic here. Does this mean you can't *resume* if the fan isn't running?

**Comment by dragomirecky** [comment]
Adding @XPila to the reviewers because of the marlin_server part

**Review by XPila — DISMISSED** [review]

**Review by dragomirecky — APPROVED** [review]

**Pull request merged**